In [1]:
#importing library
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

# **Dataset:**

In [2]:
dff = pd.read_excel('Online retail.xlsx', names=['items']) # Assuming file is in the same directory

# Assuming the separator is ';' and the column to split is 'Description'
df_split = dff['items'].str.split(',', expand=True)

df_split.head()


0          1           2                 3          4     5   \
0         burgers  meatballs        eggs              None       None  None   
1         chutney       None        None              None       None  None   
2          turkey    avocado        None              None       None  None   
3   mineral water       milk  energy bar  whole wheat rice  green tea  None   
4  low fat yogurt       None        None              None       None  None   

     6     7     8     9     10    11    12    13    14    15    16    17  \
0  None  None  None  None  None  None  None  None  None  None  None  None   
1  None  None  None  None  None  None  None  None  None  None  None  None   
2  None  None  None  None  None  None  None  None  None  None  None  None   
3  None  None  None  None  None  None  None  None  None  None  None  None   
4  None  None  None  None  None  None  None  None  None  None  None  None   

     18  
0  None  
1  None  
2  None  
3  None  
4  None

In [3]:
dataset = df_split.replace(np.nan, '', regex=True)
dataset

0                  1            2                 3   \
0            burgers          meatballs         eggs                     
1            chutney                                                     
2             turkey            avocado                                  
3      mineral water               milk   energy bar  whole wheat rice   
4     low fat yogurt                                                     
...              ...                ...          ...               ...   
7495          butter         light mayo  fresh bread                     
7496         burgers  frozen vegetables         eggs      french fries   
7497         chicken                                                     
7498        escalope          green tea                                  
7499            eggs    frozen smoothie  yogurt cake    low fat yogurt   

             4          5  6  7  8  9  10 11 12 13 14 15 16 17 18  
0                                                                  
1                                                                  
2                                                                  
3     green tea                                                    
4                                                                  
...         ...        ... .. .. .. .. .. .. .. .. .. .. .. .. ..  
7495                                                               
7496  magazines  green tea                                         
7497                                                               
7498                                                               
7499                                                               

[7500 rows x 19 columns]

In [4]:
dataset
transactions = []
for i in range(0, 7500):
    transactions.append([str(dataset.values[i,u]) for u in range(0, 19)])


In [5]:
groceries_series  = pd.DataFrame(pd.Series(transactions))
groceries_series
groceries_series.columns = ["business"]
groceries_series

business
0     [burgers, meatballs, eggs, , , , , , , , , , ,...
1         [chutney, , , , , , , , , , , , , , , , , , ]
2     [turkey, avocado, , , , , , , , , , , , , , , ...
3     [mineral water, milk, energy bar, whole wheat ...
4     [low fat yogurt, , , , , , , , , , , , , , , ,...
...                                                 ...
7495  [butter, light mayo, fresh bread, , , , , , , ...
7496  [burgers, frozen vegetables, eggs, french frie...
7497      [chicken, , , , , , , , , , , , , , , , , , ]
7498  [escalope, green tea, , , , , , , , , , , , , ...
7499  [eggs, frozen smoothie, yogurt cake, low fat y...

[7500 rows x 1 columns]

In [6]:
X = groceries_series['business'].str.join(sep=',').str.get_dummies(sep=',')
X

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0              0        0                  0          0        0            0   
1              0        0                  0          0        0            0   
2              0        0                  0          0        1            0   
3              0        0                  0          0        0            0   
4              0        0                  0          0        0            0   
...          ...      ...                ...        ...      ...          ...   
7495           0        0                  0          0        0            0   
7496           0        0                  0          0        0            0   
7497           0        0                  0          0        0            0   
7498           0        0                  0          0        0            0   
7499           0        0                  0          0        0            0   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0         0               0          0            0  ...       0   
1         0               0          0            0  ...       0   
2         0               0          0            0  ...       1   
3         0               0          0            0  ...       0   
4         0               0          0            0  ...       0   
...     ...             ...        ...          ...  ...     ...   
7495      0               0          0            0  ...       0   
7496      0               0          0            0  ...       0   
7497      0               0          0            0  ...       0   
7498      0               0          0            0  ...       0   
7499      0               0          0            0  ...       0   

      vegetables mix  water spray  white wine  whole weat flour  \
0                  0            0           0                 0   
1                  0            0           0                 0   
2                  0            0           0                 0   
3                  0            0           0                 0   
4                  0            0           0                 0   
...              ...          ...         ...               ...   
7495               0            0           0                 0   
7496               0            0           0                 0   
7497               0            0           0                 0   
7498               0            0           0                 0   
7499               0            0           0                 0   

      whole wheat pasta  whole wheat rice  yams  yogurt cake  zucchini  
0                     0                 0     0            0         0  
1                     0                 0     0            0         0  
2                     0                 0     0            0         0  
3                     0                 1     0            0         0  
4                     0                 0     0            0         0  
...                 ...               ...   ...          ...       ...  
7495                  0                 0     0            0         0  
7496                  0                 0     0            0         0  
7497                  0                 0     0            0         0  
7498                  0                 0     0            0         0  
7499                  0                 0     0            1         0  

[7500 rows x 120 columns]

# **Data Preprocessing:**

In [7]:

df=X.copy()

In [8]:

#information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500 entries, 0 to 7499
Columns: 120 entries,  asparagus to zucchini
dtypes: int64(120)
memory usage: 6.9 MB


In [9]:
#checking for null values
df.isnull().sum()

asparagus           0
almonds              0
antioxydant juice    0
asparagus            0
avocado              0
                    ..
whole wheat pasta    0
whole wheat rice     0
yams                 0
yogurt cake          0
zucchini             0
Length: 120, dtype: int64

In [10]:
#checking for duplicates
df.duplicated().sum()

2347

In [11]:
#dropping duplicate
df1=df.copy()
df1=df1.drop_duplicates()
df1

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0              0        0                  0          0        0            0   
1              0        0                  0          0        0            0   
2              0        0                  0          0        1            0   
3              0        0                  0          0        0            0   
4              0        0                  0          0        0            0   
...          ...      ...                ...        ...      ...          ...   
7492           0        0                  0          0        0            0   
7493           0        0                  0          0        0            0   
7495           0        0                  0          0        0            0   
7496           0        0                  0          0        0            0   
7499           0        0                  0          0        0            0   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0         0               0          0            0  ...       0   
1         0               0          0            0  ...       0   
2         0               0          0            0  ...       1   
3         0               0          0            0  ...       0   
4         0               0          0            0  ...       0   
...     ...             ...        ...          ...  ...     ...   
7492      0               0          0            0  ...       0   
7493      0               0          0            0  ...       1   
7495      0               0          0            0  ...       0   
7496      0               0          0            0  ...       0   
7499      0               0          0            0  ...       0   

      vegetables mix  water spray  white wine  whole weat flour  \
0                  0            0           0                 0   
1                  0            0           0                 0   
2                  0            0           0                 0   
3                  0            0           0                 0   
4                  0            0           0                 0   
...              ...          ...         ...               ...   
7492               0            0           0                 0   
7493               0            0           0                 0   
7495               0            0           0                 0   
7496               0            0           0                 0   
7499               0            0           0                 0   

      whole wheat pasta  whole wheat rice  yams  yogurt cake  zucchini  
0                     0                 0     0            0         0  
1                     0                 0     0            0         0  
2                     0                 0     0            0         0  
3                     0                 1     0            0         0  
4                     0                 0     0            0         0  
...                 ...               ...   ...          ...       ...  
7492                  0                 0     0            0         0  
7493                  0                 1     0            0         0  
7495                  0                 0     0            0         0  
7496                  0                 0     0            0         0  
7499                  0                 0     0            1         0  

[5153 rows x 120 columns]

In [12]:
#creating dummies
df1=pd.get_dummies(df1)
df1= df1.astype(int)
df1

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0              0        0                  0          0        0            0   
1              0        0                  0          0        0            0   
2              0        0                  0          0        1            0   
3              0        0                  0          0        0            0   
4              0        0                  0          0        0            0   
...          ...      ...                ...        ...      ...          ...   
7492           0        0                  0          0        0            0   
7493           0        0                  0          0        0            0   
7495           0        0                  0          0        0            0   
7496           0        0                  0          0        0            0   
7499           0        0                  0          0        0            0   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0         0               0          0            0  ...       0   
1         0               0          0            0  ...       0   
2         0               0          0            0  ...       1   
3         0               0          0            0  ...       0   
4         0               0          0            0  ...       0   
...     ...             ...        ...          ...  ...     ...   
7492      0               0          0            0  ...       0   
7493      0               0          0            0  ...       1   
7495      0               0          0            0  ...       0   
7496      0               0          0            0  ...       0   
7499      0               0          0            0  ...       0   

      vegetables mix  water spray  white wine  whole weat flour  \
0                  0            0           0                 0   
1                  0            0           0                 0   
2                  0            0           0                 0   
3                  0            0           0                 0   
4                  0            0           0                 0   
...              ...          ...         ...               ...   
7492               0            0           0                 0   
7493               0            0           0                 0   
7495               0            0           0                 0   
7496               0            0           0                 0   
7499               0            0           0                 0   

      whole wheat pasta  whole wheat rice  yams  yogurt cake  zucchini  
0                     0                 0     0            0         0  
1                     0                 0     0            0         0  
2                     0                 0     0            0         0  
3                     0                 1     0            0         0  
4                     0                 0     0            0         0  
...                 ...               ...   ...          ...       ...  
7492                  0                 0     0            0         0  
7493                  0                 1     0            0         0  
7495                  0                 0     0            0         0  
7496                  0                 0     0            0         0  
7499                  0                 0     0            1         0  

[5153 rows x 120 columns]

# **Apriori Algorithm**

In [13]:
# Applying Apriori Algorithm
from mlxtend.frequent_patterns import apriori,association_rules
frequent_itemsets = apriori(df1, min_support=0.01,max_len=4,use_colnames = True)
frequent_itemsets

/usr/local/lib/python3.11/dist-packages/mlxtend/frequent_patterns/fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                               itemsets
0    0.029303                              (almonds)
1    0.011062                    (antioxydant juice)
2    0.045993                              (avocado)
3    0.012614                                (bacon)
4    0.015525                       (barbecue sauce)
..        ...                                    ...
431  0.014749  (spaghetti, olive oil, mineral water)
432  0.016689   (spaghetti, pancakes, mineral water)
433  0.012420     (spaghetti, shrimp, mineral water)
434  0.010867       (soup, spaghetti, mineral water)
435  0.013584   (tomatoes, spaghetti, mineral water)

[436 rows x 2 columns]

In [14]:
#finding according to lift
# Calculate the number of itemsets in frequent_itemsets
num_itemsets = len(frequent_itemsets)

# Pass num_itemsets to association_rules
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1, num_itemsets=num_itemsets)

rules.head(20)
rules.sort_values('lift', ascending=False).head(10)


antecedents                     consequents  \
428             (whole wheat pasta)                     (olive oil)   
429                     (olive oil)             (whole wheat pasta)   
798           (milk, mineral water)                          (soup)   
799                          (soup)           (milk, mineral water)   
302                   (ground beef)                 (herb & pepper)   
303                 (herb & pepper)                   (ground beef)   
739         (shrimp, mineral water)             (frozen vegetables)   
742             (frozen vegetables)         (shrimp, mineral water)   
722  (frozen vegetables, spaghetti)                   (ground beef)   
723                   (ground beef)  (frozen vegetables, spaghetti)   

     antecedent support  consequent support   support  confidence      lift  \
428            0.040753            0.087910  0.011062    0.271429  3.087575   
429            0.087910            0.040753  0.011062    0.125828  3.087575   
798            0.067922            0.071221  0.012420    0.182857  2.567474   
799            0.071221            0.067922  0.012420    0.174387  2.567474   
302            0.136425            0.066757  0.022899    0.167852  2.514365   
303            0.066757            0.136425  0.022899    0.343023  2.514365   
739            0.033573            0.130409  0.010479    0.312139  2.393528   
742            0.130409            0.033573  0.010479    0.080357  2.393528   
722            0.039200            0.136425  0.012614    0.321782  2.358668   
723            0.136425            0.039200  0.012614    0.092461  2.358668   

     representativity  leverage  conviction  zhangs_metric   jaccard  \
428               1.0  0.007479    1.251888       0.704846  0.094059   
429               1.0  0.007479    1.097320       0.741288  0.094059   
798               1.0  0.007583    1.136618       0.655001  0.098009   
799               1.0  0.007583    1.128953       0.657327  0.098009   
302               1.0  0.013792    1.121487       0.697433  0.127018   
303               1.0  0.013792    1.314468       0.645368  0.127018   
739               1.0  0.006101    1.264195       0.602432  0.068268   
742               1.0  0.006101    1.050872       0.669518  0.068268   
722               1.0  0.007266    1.273300       0.599534  0.077381   
723               1.0  0.007266    1.058687       0.667032  0.077381   

     certainty  kulczynski  
428   0.201207    0.198628  
429   0.088689    0.198628  
798   0.120197    0.178622  
799   0.114224    0.178622  
302   0.108326    0.255438  
303   0.239236    0.255438  
739   0.208983    0.196248  
742   0.048410    0.196248  
722   0.214639    0.207122  
723   0.055433    0.207122

In [15]:
# High chances of if and then
rules.sort_values('lift',ascending = False)[0:20]


antecedents                       consequents  \
428               (whole wheat pasta)                       (olive oil)   
429                       (olive oil)               (whole wheat pasta)   
798             (milk, mineral water)                            (soup)   
799                            (soup)             (milk, mineral water)   
302                     (ground beef)                   (herb & pepper)   
303                   (herb & pepper)                     (ground beef)   
739           (shrimp, mineral water)               (frozen vegetables)   
742               (frozen vegetables)           (shrimp, mineral water)   
722    (frozen vegetables, spaghetti)                     (ground beef)   
723                     (ground beef)    (frozen vegetables, spaghetti)   
809                 (milk, spaghetti)                       (olive oil)   
812                       (olive oil)                 (milk, spaghetti)   
720  (ground beef, frozen vegetables)                       (spaghetti)   
725                       (spaghetti)  (ground beef, frozen vegetables)   
800                            (milk)             (mineral water, soup)   
797             (mineral water, soup)                            (milk)   
466              (french fries, eggs)                         (burgers)   
471                         (burgers)              (french fries, eggs)   
774        (mineral water, spaghetti)                     (ground beef)   
775                     (ground beef)        (mineral water, spaghetti)   

     antecedent support  consequent support   support  confidence      lift  \
428            0.040753            0.087910  0.011062    0.271429  3.087575   
429            0.087910            0.040753  0.011062    0.125828  3.087575   
798            0.067922            0.071221  0.012420    0.182857  2.567474   
799            0.071221            0.067922  0.012420    0.174387  2.567474   
302            0.136425            0.066757  0.022899    0.167852  2.514365   
303            0.066757            0.136425  0.022899    0.343023  2.514365   
739            0.033573            0.130409  0.010479    0.312139  2.393528   
742            0.130409            0.033573  0.010479    0.080357  2.393528   
722            0.039200            0.136425  0.012614    0.321782  2.358668   
723            0.136425            0.039200  0.012614    0.092461  2.358668   
809            0.050262            0.087910  0.010285    0.204633  2.327759   
812            0.087910            0.050262  0.010285    0.116998  2.327759   
720            0.024646            0.230157  0.012614    0.511811  2.223746   
725            0.230157            0.024646  0.012614    0.054806  2.223746   
800            0.170386            0.033573  0.012420    0.072893  2.171198   
797            0.033573            0.170386  0.012420    0.369942  2.171198   
466            0.046575            0.114302  0.011450    0.245833  2.150729   
471            0.114302            0.046575  0.011450    0.100170  2.150729   
774            0.085193            0.136425  0.024840    0.291572  2.137225   
775            0.136425            0.085193  0.024840    0.182077  2.137225   

     representativity  leverage  conviction  zhangs_metric   jaccard  \
428               1.0  0.007479    1.251888       0.704846  0.094059   
429               1.0  0.007479    1.097320       0.741288  0.094059   
798               1.0  0.007583    1.136618       0.655001  0.098009   
799               1.0  0.007583    1.128953       0.657327  0.098009   
302               1.0  0.013792    1.121487       0.697433  0.127018   
303               1.0  0.013792    1.314468       0.645368  0.127018   
739               1.0  0.006101    1.264195       0.602432  0.068268   
742               1.0  0.006101    1.050872       0.669518  0.068268   
722               1.0  0.007266    1.273300       0.599534  0.077381   
723               1.0  0.007266    1.058687       0.667032  0.077381  

In [16]:

#Highest chances of lift of rare items
rules[rules.lift>3]

antecedents          consequents  antecedent support  \
428  (whole wheat pasta)          (olive oil)            0.040753   
429          (olive oil)  (whole wheat pasta)            0.087910   

     consequent support   support  confidence      lift  representativity  \
428            0.087910  0.011062    0.271429  3.087575               1.0   
429            0.040753  0.011062    0.125828  3.087575               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
428  0.007479    1.251888       0.704846  0.094059   0.201207    0.198628  
429  0.007479    1.097320       0.741288  0.094059   0.088689    0.198628

In [17]:
#these are likely to buy together
# Calculate the number of itemsets in frequent_itemsets
num_itemsets = len(frequent_itemsets)

# Pass num_itemsets to association_rules
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.5, num_itemsets=num_itemsets)

rules.head(20)
rules.sort_values('confidence', ascending=False).head(10)

antecedents      consequents  antecedent support  \
8                      (milk, soup)  (mineral water)            0.021541   
3  (ground beef, frozen vegetables)  (mineral water)            0.024646   
9                 (spaghetti, soup)  (mineral water)            0.020765   
6           (ground beef, pancakes)  (mineral water)            0.020959   
0              (chicken, chocolate)  (mineral water)            0.021153   
7                 (milk, olive oil)  (mineral water)            0.024258   
4  (ground beef, frozen vegetables)      (spaghetti)            0.024646   
1            (olive oil, chocolate)  (mineral water)            0.023676   
5               (ground beef, milk)  (mineral water)            0.031826   
2               (eggs, ground beef)  (mineral water)            0.028915   

   consequent support   support  confidence      lift  representativity  \
8            0.299825  0.012420    0.576577  1.923041               1.0   
3            0.299825  0.013390    0.543307  1.812079               1.0   
9            0.299825  0.010867    0.523364  1.745565               1.0   
6            0.299825  0.010867    0.518519  1.729402               1.0   
0            0.299825  0.010867    0.513761  1.713536               1.0   
7            0.299825  0.012420    0.512000  1.707661               1.0   
4            0.230157  0.012614    0.511811  2.223746               1.0   
1            0.299825  0.012032    0.508197  1.694976               1.0   
5            0.299825  0.016107    0.506098  1.687975               1.0   
2            0.299825  0.014555    0.503356  1.678830               1.0   

   leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
8  0.005961    1.653604       0.490557  0.040201   0.395260    0.309000  
3  0.006001    1.533141       0.459472  0.043044   0.347744    0.293984  
9  0.004642    1.468994       0.436176  0.035088   0.319262    0.279805  
6  0.004584    1.454209       0.430794  0.035066   0.312341    0.277382  
0  0.004525    1.439982       0.425410  0.035044   0.305547    0.275004  
7  0.005147    1.434784       0.424706  0.039851   0.303031    0.276712  
4  0.006942    1.576936       0.564214  0.052083   0.365859    0.283309  
1  0.004933    1.423688       0.419964  0.038629   0.297599    0.274163  
5  0.006565    1.417638       0.420972  0.051046   0.294601    0.279910  
2  0.005885    1.409811       0.416387  0.046325   0.290685    0.275950

# **Analysis and Interpretation:**

Key Insights: According to lift

1.Strong Associations: There are several rules with high lift values, such as the association between "olive oil" and "whole wheat pasta" (lift = 3.09), indicating strong relationships.

2.High Confidence: The rule involving "herb & pepper" → "ground beef" shows high confidence (34.3%), suggesting that herb & pepper is frequently purchased with ground beef.

3.Item Pairings: Some rules show that pairs of items like "soup" and "mineral water" often lead to the purchase of other items like "milk." These pairs are valuable for bundling products or creating targeted promotions.

4.Lift and Support Relationship: Several rules with higher support (e.g., "spaghetti, mineral water" → "ground beef") show lower lift, suggesting that although the antecedents and consequents are popular, their joint occurrence may not be as impactful.

5.Utilizing Data for Marketing: The dataset reveals opportunities for cross-promotions or bundle offers based on product relationships. For instance, offering "whole wheat pasta" as a complementary product when customers buy "olive oil" could be a strategic marketing move. ''' ''' Key Insights: According to confidence

1.Soup, Milk → Mineral Water (0.5766): High confidence, indicating strong likelihood of customers buying mineral water with soup and milk.

2.Frozen Vegetables, Ground Beef → Mineral Water (0.5433): Moderate confidence, suggesting a good chance of purchasing mineral water with frozen vegetables and ground beef.

3.Soup, Spaghetti → Mineral Water (0.5234): Solid confidence, showing customers tend to buy mineral water when purchasing soup and spaghetti.

4.Pancakes, Ground Beef → Mineral Water (0.5185): Fair confidence, indicating mineral water is likely to be bought with pancakes and ground beef.

5.Chicken, Chocolate → Mineral Water (0.5138): Decent confidence, showing a reasonable chance of buying mineral water alongside chicken and chocolate.

# **INTERVIEW QUESTIONS**

1.What is Lift and Why is it Important?
Lift measures how much more likely the consequent occurs given the antecedent, compared to random chance (Lift=Confidence/Support of Consequent).

It identifies meaningful rules by filtering out those based solely on high support items.

A lift > 1 indicates a positive association, making it crucial for finding significant relationships.

2.What are Support and Confidence, and How Are They Calculated?
Support: The fraction of transactions containing an itemset (Support=Transactions with Itemset/Total Transactions).

Confidence: The likelihood of the consequent given the antecedent (Confidence= Support of Antecedent/Support of Rule).

Both measure the frequency and reliability of association rules in the dataset.

3.What Are Some Limitations of Association Rule Mining?
It can generate too many rules, making it hard to identify meaningful patterns.

Rare but significant associations might be missed due to strict support thresholds.

High computational cost for large datasets or itemsets with many combinations.